pkg/cluster/deploy.go

```diff
@@ -30,3 +45,117 @@ func (u *simpleUpgrader) Deploy(ctx context.Context, cs *api.OpenShiftManagedClu
 	}
 	return nil
 }
+
+type scalesetName string
+type instanceName string
+type hash string
+
+type vmInfo struct {
+	ComputerName instanceName `json:"computerName,omitempty"`
+	ScalesetHash hash         `json:"scalesetHash,omitempty"`
+}
+
+func hashScaleSets(azuretemplate map[string]interface{}) (map[scalesetName]hash, error) {
+	ssHashes := make(map[scalesetName]hash)
+	for _, r := range jsonpath.MustCompile("$.resources[?(@.type='Microsoft.Compute/virtualMachineScaleSets')]").Get(azuretemplate) {
+		original, ok := r.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		// deep-copy the ARM template since we are mutating it below.
+		resource := deepCopy(original)
+
+		// cleanup capacity so that no unnecessary VM rotations are going
+		// to occur because of a scale up/down.
+		jsonpath.MustCompile("$.sku.capacity").Delete(resource)
+
+		// filter out the nsg dependsOn entry since we remove it
+		// during upgrades due to an azure issue.
+		jsonpath.MustCompile("$.dependsOn").Delete(resource)
+
+		// hash scale set
+		data, err := json.Marshal(resource)
+		if err != nil {
+			return nil, err
+		}
+		hf := sha256.New()
+		hf.Write(data)
+
+		scaleSetName := jsonpath.MustCompile("$.name").MustGetString(resource)
+		ssHashes[scalesetName(scaleSetName)] = hash(base64.StdEncoding.EncodeToString(hf.Sum(nil)))
+	}
+	return ssHashes, nil
+}
+
+func deepCopy(in map[string]interface{}) map[string]interface{} {
+	b, err := json.Marshal(in)
+	if err != nil {
+		panic(err)
+	}
+	var out map[string]interface{}
+	err = json.Unmarshal(b, &out)
+	if err != nil {
+		panic(err)
+	}
+	return out
+}
+
+func (u *simpleUpgrader) initializeUpdateBlob(cs *api.OpenShiftManagedCluster, ssHashes map[scalesetName]hash) error {
+	vmHashes := make(map[instanceName]hash)
+	for _, profile := range cs.Properties.AgentPoolProfiles {
+		for i := 0; i < profile.Count; i++ {
+			name := instanceName(fmt.Sprintf("ss-%s_%d", profile.Name, i))
+			vmHashes[name] = ssHashes[scalesetName("ss-"+profile.Name)]
+		}
+	}
+	return u.updateBlob(vmHashes)
+}
+
+const updateContainerName = "update"
+const updateBlobName = "update"
+
+func (u *simpleUpgrader) updateBlob(b map[instanceName]hash) error {
```

Comment:
open follow up for mock tests here

Reply:
https://github.com/openshift/openshift-azure/issues/737